lib/kafka/async_producer.rb

```diff
@@ -0,0 +1,117 @@
+require "thread"
+
+module Kafka
+
+  # A Kafka producer that does all its work in the background so as to not block
+  # the calling thread. Calls to {#deliver_messages} are asynchronous and return
+  # immediately.
+  #
+  # In addition to this property it's possible to define automatic delivery
+  # policies. These allow placing an upper bound on the number of buffered
+  # messages and the time between message deliveries.
+  #
+  # * If `delivery_threshold` is set to a value _n_ higher than zero, the producer
+  #   will automatically deliver its messages once its buffer size reaches _n_.
+  # * If `delivery_interval` is set to a value _n_ higher than zero, the producer
+  #   will automatically deliver its messages every _n_ seconds.
+  #
+  # ## Example
+  #
+  #     producer = kafka.get_async_producer(
+  #       # Keep at most 1.000 messages in the buffer before delivering:
+  #       delivery_threshold: 1000,
+  #
+  #       # Deliver messages every 30 seconds:
+  #       delivery_interval: 30,
+  #     )
+  #
+  #     # There's no need to manually call #deliver_messages, it will happen
+  #     # automatically in the background.
+  #     producer.produce("hello", topic: "greetings")
+  #
+  class AsyncProducer
+    def initialize(sync_producer:, delivery_threshold: 0, delivery_interval:)
+      @queue = Queue.new
+
+      @worker_thread = Thread.new do
+        worker = Worker.new(
+          queue: @queue,
+          producer: sync_producer,
+          delivery_threshold: delivery_threshold,
+        )
+
+        worker.run
+      end
+
+      @worker_thread.abort_on_exception = true
+
+      if delivery_interval > 0
+        Thread.new do
+          Timer.new(queue: @queue, interval: delivery_interval).run
+        end
+      end
+    end
+
+    def produce(*args)
+      @queue << [:produce, args]
+    end
+
+    def deliver_messages
+      @queue << [:deliver_messages, nil]
+    end
+
+    def shutdown
+      @queue << [:shutdown, nil]
+      @worker_thread.join
+    end
+
+    class Timer
+      def initialize(interval:, queue:)
+        @queue = queue
+        @interval = interval
+      end
+
+      def run
+        loop do
+          sleep(@interval)
+          @queue << [:deliver_messages, nil]
+        end
+      end
+    end
+
+    class Worker
+      def initialize(queue:, producer:, delivery_threshold:)
+        @queue = queue
+        @producer = producer
+        @delivery_threshold = delivery_threshold
+      end
+
+      def run
+        loop do
+          operation, payload = @queue.pop
+
+          case operation
+          when :produce
+            @producer.produce(*payload)
+            @producer.deliver_messages if threshold_reached?
+          when :deliver_messages
+            @producer.deliver_messages
+          when :shutdown
```

Comment:
should this call a last `deliver_messages` before shutting down? Guess we could add a `shutdown_now` or something that doesn't do that?